### ISSUE TITLE
Non-str `Secret` fields are always str on `json` serialization

### ISSUE THREAD

**Issue by Danipulok** [pydantic/pydantic#9827]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Current behavior for non-str `Secret` types is unexpected in my opinion.
Example code:
```python
from pydantic import BaseModel, Secret


class Foo(BaseModel):
    secret_data: Secret[dict]

foo = Foo(secret_data={"bar": "bar_value"})
dumped_jsonlike = foo.model_dump(mode='json')
print(dumped_jsonlike)
#> {'secret_data': '**********'}

Foo.model_validate(dumped_jsonlike)
"""
pydantic_core._pydantic_core.ValidationError: 1 validation error for Foo
secret_data
  Input should be a valid dictionary [type=dict_type, input_value='**********', input_type=str]
    For further information visit https://errors.pydantic.dev/2.8/v/dict_type
"""
```

In our project code we have to serialize and then load back some models with `Secret` fields.
When fields are secret, we don't need them dumped and don't care about them, so it would be really convenient if dumped type is still as original proper loading and being correspond to the scheme.

### 1. Add `__bool__` method to `Secret` types that would call `__bool__` for its secret value.
Notes: 
1. Currently similar logic is already used for [_display](https://github.com/pydantic/pydantic/blob/2b7b50dc47a86eb294f6d03c04387ff567caed48/pydantic/types.py#L1540), so I think it will only add convenience.
2. This was implemented in my PR: [link](https://github.com/pydantic/pydantic/pull/9819/files#diff-f7a381c0918da7ec905bfbe6aff2ccc72c2d0fd6da0dd3011ef1da892e42d57aR1481-R1483)

### 2. Change `__repr__` method for `Secret` types.
When currently calling `repr` on `Secret` fields, it's impossible to determine a type.
**Current behavior**:
```python
from pydantic import BaseModel, Secret


class Foo(BaseModel):
    secret_data: Secret[dict]

foo = Foo(secret_data={"bar": "bar_value"})
print(repr(foo))
#> Foo(secret_data=Secret('**********'))
```

**Proposed behavior**:
```python
from pydantic import BaseModel, Secret


class Foo(BaseModel):
    secret_data: Secret[dict]

foo = Foo(secret_data={"bar": "bar_value"})
print(repr(foo))
#> Foo(secret_data=Secret(type=dict, filled=True))
```

Advantages:
1. We know the type of secret field without need to look for model and checking field type.

Notes: 
1. I'm not sure about the naming of `filled`, maybe `is_set`, `is_filled` is better.
2. If type is custom type, return `__class__.__name__`, example: `Foo`.
3. `__str__` behavior is not changed and will still call `self._display`
4. If this is accepted please specify if `repr` for `SecretStr` and `SecretBytes` should stay the same or be changed too.

### 3. Change serialization of `Secret` to have the same type when dumping to json.
On serialization, we serialize default value for this type. 
For example:
- `dict`, model -> `'{}'`;
- `set`, `list`, `tuple` -> `'[]'`;
- `int` -> `'0'`;
- `bool` -> `'false'`.

**Current behavior**:
```python
from pydantic import BaseModel, Secret


class Foo(BaseModel):
    secret_data: Secret[dict]

foo = Foo(secret_data={"bar": "bar_value"})
dumped_jsonlike = foo.model_dump(mode='json')
print(dumped_jsonlike)
#> {'secret_data': '**********'}
```

**Proposed behavior**:
```python
from pydantic import BaseModel, Secret


class Foo(BaseModel):
    secret_data: Secret[dict]

foo = Foo(secret_data={"bar": "bar_value"})
dumped_jsonlike = foo.model_dump(mode='json')
print(dumped_jsonlike)
#> {'secret_data': '{}'}
```

Advantages:
1. Keep dumped data as expected by `JsonSchema`. Currently it's exported to `'type': 'object'`, which is valid and expected.
2. Add opportunity to call `Model.model_validate` for dumped data.
3. Allow proper dumping and having better user experience with  

Notes:
1. We keep the dumped type the same: `str`. Meaning it's not such breaking change as it may seems at first.
2. In this paragraph I propose changing specifically `Secret` behavior. Not `SecretStr` and `SecretBytes`.
3. This was implemented in my PR: [link](https://github.com/pydantic/pydantic/pull/9819/files#diff-b6337d7bb9ba2ed222ac52055eefb9f98223428d82337d552e6dd3e0a91cfb49R59-R63)

### 4. Change serialization of `SecretStr` and `SecretBytes` to have the same type when dumping to json.
Basically all the same as in paragraph 3.
But since those types have been in the library for quite some time I suppose you wouldn't like for their serialization to be changed.

### Python, Pydantic & OS Version

```Text
pydantic version: 2.8.0
        pydantic-core version: 2.20.0
          pydantic-core build: profile=release pgo=true
                 install path: D:\Code\Aliyoop\pydantic\pydantic
               python version: 3.11.8 (tags/v3.11.8:db85d51, Feb  6 2024, 22:03:32) [MSC v.1937 64 bit (AMD64)]
                     platform: Windows-10-10.0.19045-SP0
             related packages: mypy-1.10.1 typing_extensions-4.12.2
                       commit: 28aedcfb
```

**Comment by Danipulok** [comment]
This it mostly related to my PR #9819 which I did without discussing with you what should and should not be changed and how (I'm sorry once again).

I would like to discuss all points specified in the issue, because our team has decided to use `pydantic` secrets and would like to make the experience better for us and everyone else!

Ready to discuss
Since @sydney-runkle is related to my previous PR, I also would like to tag her
Thank you all in advance!

**Comment by Danipulok** [comment]
@sydney-runkle good day!
Any updates or maybe decisions related to this?

**Comment by sydney-runkle** [comment]
@Danipulok,

Thanks for the ping - just got back from holiday today. I'll bring this up with the team at our next open source sync and get back to you after that :).

**Comment by Danipulok** [comment]
Hey @sydney-runkle !

Any news on the topic?

We can leave `SecretStr` and `SecretBytes` as now since they are popular, but maybe change `Secret`?

**Comment by sydney-runkle** [comment]
@Danipulok,

Hey! Let's move forward with some of these changes. That being said, I don't think we can change the default behavior (don't want to break things for existing users). Happy to add support for more intuitive serialization though :).

Thanks so much for your patience. Looking forward to working with you on this!

**Comment by Danipulok** [comment]
@sydney-runkle, thanks for the reply, glad to hear!

And what are exactly the changes you would accept? What paragraphs from the issue description?

Maybe 1-3 to change only Secret behaviour but keep SecretStr and SecretBytes?

**Comment by Danipulok** [comment]
@sydney-runkle hey again!

Could you please specify with what parts of the initial issue/PR should I work with at new PR?

I would really like to implement something and make `Secret`s better

**Comment by sydney-runkle** [comment]
> 1. Add __bool__ method to Secret types that would call __bool__ for its secret value.

I think this one will have to remain in user code - it's a breaking change for folks to have **more** information available on their secret instance due to a change in a minor release.

> 2. Change __repr__ method for Secret types.

By default, we need to keep the current behavior for `__repr__` and `_display` for backwards compatibility reasons. Folks are able to customize the result of the `_display` function to their hearts' content depending on the custom `Secret` type. So, not sure what else needs to be done here, given that customization is endorsed and documented.

> 3. Change serialization of Secret to have the same type when dumping to json.

It makes sense to me that you'd want to change this behavior. One current way to do this is to use a custom serializer, like:

```py
from pydantic import BaseModel, Secret, PlainSerializer
from typing_extensions import Annotated


# Overwriting the representation
class SecretSalary(Secret[float]):
    def _display(self) -> str:
        return '$****.**'


class Employee(BaseModel):
    salary: Annotated[SecretSalary, PlainSerializer(lambda _: 0)]


employee = Employee(salary=42)

print(employee)
#> salary=SecretSalary('$****.**')

print(employee.model_dump())
#> {'salary': 0}
```

Perhaps we could provide some sort of hook for custom secret types that allows easy customization of their serialization logic. In the short term though, this annotated serializer approach works pretty well.

Maybe we could just document this as an expected pattern?

> 4. Change serialization of SecretStr and SecretBytes to have the same type when dumping to json.

Indeed, we can't change this.

Looking forward to hearing your thoughts.

**Comment by sydney-runkle** [comment]
@Danipulok, thoughts on the above?

**Comment by Danipulok** [comment]
@sydney-runkle , hi!
Sorry for now responding earlier

Unfortunately, I don't think I can think of a solution that in my opinion would be valid (like passing the validation again) without changing the default behavior.

I can think only of four ways in total to do this:
1. Defining every type of a needed secret value in user code as you have specified in your example:
```python
from pydantic import BaseModel, model_serializer, Secret

class SecretSalary(Secret[float]):
    def _display(self) -> str:
        return '$****.**'

    @model_serializer()
    def __model_serialize(self) -> str:
        return self._display()

class Worker(BaseModel):
    name: str
    salary: SecretSalary

worker = Worker(name="John", salary=1234.56)
print(worker.model_dump(mode="json"))  # {'name': 'John', 'salary': '$****.**'}
```

2. Have special config for secret fields, for example:
```python
from pydantic import BaseModel, Field, model_serializer, Secret

class Worker(BaseModel):
    name: str
    salary: Secret[float] = Field(serialization_value=0)

worker = Worker(name="John", salary=1234.56)
print(worker.model_dump(mode="json"))  # {'name': 'John', 'salary': 0}
```

3. Change the default `Secret` behavior:
```python
from pydantic import BaseModel, Secret

class Worker(BaseModel):
    name: str
    salary: Secret[float]

worker = Worker(name="John", salary=1234.56)
print(worker.model_dump(mode="json"))  # {'name': 'John', 'salary': 0}
```

4. Combine your option and do not create any new entities in `pydantic`, but create a `SecretField` **function** in `pydantic`, that works like this:
```python
from typing import Any, TypeVar
from pydantic import BaseModel, model_serializer, Secret

SecretT = TypeVar("SecretT", bound=type)

def SecretField(
    type_: SecretT,
    *,
    serialization_value: Any = None,
    display_value: Any = None,
) -> type[Secret[SecretT]]:
    class SecretItem(Secret[type_]):
        def _display(self) -> str:
            return display_value if display_value is not None else super()._display()

        @model_serializer()
        def __model_serialize(self) -> Any:
            return serialization_value if serialization_value is not None else self._display()

    return SecretItem

SecretSalary = SecretField(
    float,
    serialization_value=0,
    display_value='$****.**',
)

class Worker(BaseModel):
    name: str
    salary: SecretSalary

worker = Worker(name="John", salary=1234.56)
print(worker.model_dump(mode="json"))  # 
```

I also noticed that option1 and option4 don't actually work because they need `Annotated`, as you have specified in your example.
So I guess it would be good if at least private method `_serialize` it added to `Secret` and by default it calls `self._display`.
At least it would allow having something like option4 on user side with easier user experience
